src/credentials/CHIPCert.cpp

```diff
@@ -708,22 +708,109 @@ CHIP_ERROR DetermineCertType(ChipCertificateData & cert)
 
 bool ChipDN::IsEqual(const ChipDN & other) const
 {
-    if (mAttrOID == kOID_Unknown || mAttrOID == kOID_NotSpecified || mAttrOID != other.mAttrOID)
+
+    if (mAttrOID[0] == chip::ASN1::kOID_AttributeType_ChipNodeId)
+        return other.mAttrOID[0] == chip::ASN1::kOID_AttributeType_ChipNodeId &&
+            mAttrValue[0].mChipId == other.mAttrValue[0].mChipId;
+
+    for (int i = 1; i < RDN_NUM; ++i)
     {
-        return false;
+        if (mAttrOID[i] == kOID_Unknown || mAttrOID[i] == kOID_NotSpecified)
+            continue;
+
+        int j = 1;
+        for (; j < RDN_NUM; ++j)
+        {
+            if (mAttrOID[i] == other.mAttrOID[j])
+                break;
+        }
+
+        if (j == RDN_NUM)
+            continue;
+
+        if (IsChipX509Attr(mAttrOID[i]))
+        {
+            if (mAttrValue[i].mChipId != other.mAttrValue[j].mChipId)
+                return false;
+        }
+        else
+        {
+            if (mAttrValue[i].mString.mLen != other.mAttrValue[j].mString.mLen ||
+                memcmp(mAttrValue[i].mString.mValue, other.mAttrValue[j].mString.mValue, mAttrValue[i].mString.mLen) != 0)
+                return false;
+        }
     }
+    return true;
+}
 
-    if (IsChipIdX509Attr(mAttrOID))
+bool ChipDN::IsEmpty() const
+{
+    int i = 0;
+
+    for (; i < RDN_NUM; ++i)
     {
-        return mAttrValue.mChipId == other.mAttrValue.mChipId;
+        if (mAttrOID[i] != chip::ASN1::kOID_NotSpecified)
+        {
+            return false;
+        }
     }
-    else
+    return true;
+}
+
+void ChipDN::Clear()
+{
+    int i = 0;
+
+    for (; i < RDN_NUM; ++i)
     {
-        return (mAttrValue.mString.mLen == other.mAttrValue.mString.mLen &&
-                memcmp(mAttrValue.mString.mValue, other.mAttrValue.mString.mValue, mAttrValue.mString.mLen) == 0);
+        mAttrOID[i] = chip::ASN1::kOID_NotSpecified;
     }
 }
 
+void ChipDN::Add(chip::ASN1::OID attrOID, uint64_t chipId)
```

Comment:
I would suggest that the add methods should surface errors when add is unsuccessful by returning a CHIP_ERROR to the caller.

Reply:
thanks, fixed

Reply:
The other piece of this is that callers of `Add()` also need to check for success.

Reply:
sure. done